WordPress/wordpress-develop
# 6.4 Backport: Add new wp patterns package

## Description

Adds the newly created @wordpress/patterns package. Currently there is duplicated code for adding patterns in post editor (imported from reusable-blocks) and site editor (local modal component), which means any changes need to be applied in two places, so the patterns package was added in order to allow sharing of this code, and also as a means to start slowly deprecating the @wordpress/reusable-blocks package

Trac ticket: https://core.trac.wordpress.org/ticket/59380

## Testing Instructions

**This can't be properly tested until the latest GB packages are backported.**

- In the post editor add a block and use the `Create pattern` block menu to create both a synced and an unsynced pattern and make sure both options work as expected
- In the post editor make sure the Detach pattern and Manage patterns options on the block overflow menu work correctly
- In the site editor pattern section use the + at the top of the left navigation to create both a synced and unsynced pattern and make sure both are created as expected
- In the site editor block editor use the `Create pattern` block menu to create both a synced and an unsynced pattern and make sure both options work as expected and also make sure the Detach pattern and Manage patterns options on the block overflow menu work correctly

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/d5d7b369-5369-48e3-801c-95d3bdd615f3

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**
